src/logic/map_objects/tribes/productionsite.cc

```diff
@@ -605,37 +599,92 @@ bool ProductionSite::warp_worker(EditorGameBase& egbase, const WorkerDescr& wdes
 	return false;
 }
 
+/**
+ * Locate the worker slot at which the given worker is currently employed.
+ * @return Index in descr().working_positions() (slot requirement) and working_positions_ (current
+ *         slot), or (-1, -1) if the worker does not work here.
+ */
+std::pair<int32_t, int32_t> ProductionSite::find_worker(OPtr<Worker> w) {
+	auto wp = working_positions_.begin();
+	const BillOfMaterials& bom = descr().working_positions();
+	int32_t wp_index = 0;
+	int32_t bom_index = 0;
+	for (auto temp_wp = bom.begin(); temp_wp != bom.end(); ++temp_wp, ++bom_index) {
+		for (uint32_t j = temp_wp->second; j; --j, ++wp, ++wp_index) {
+			if (wp->worker == w) {
+				return std::pair<int32_t, int32_t>(bom_index, wp_index);
+			}
+		}
+	}
+	return std::pair<int32_t, int32_t>(-1, -1);
+}
+
+/**
+ * Tries to replace a new worker request with a worker already present on the site
+ * @param game: Game reference to retrieve description info
+ * @param worker_index: The required worker type
+ * @param wp: Pointer to the slot to be filled (contains the worker request)
+ */
+void ProductionSite::try_replace_worker(const Game* game,
+                                        DescriptionIndex worker_index,
+                                        WorkingPosition* wp) {
+	// Search replacement worker in the same building
+	for (auto wp_repl = working_positions_.begin(); wp_repl != working_positions_.end(); wp_repl++) {
```

Comment:
```suggestion
	for (auto wp_repl = working_positions_.begin(); wp_repl != working_positions_.end(); ++wp_repl) {
```
pre-increment is more efficient than post-increment for non-trivial types